indimail-mta-x/dnstlsarr.1
.TH dnstlsarr 1

.SH NAME
dnstlsarr - DANE/TLSA RR Tester. Display TLSA RR record for a host

.SH SYNOPSIS
.B dnstlsarr
[-\fBv\fR level]
[-m]
[-s]
.I host

.SH DESCRIPTION
dnstlsarr fetches and display the TLSA RR records for \fIhost\fR. With the -s option, it can
perform a DANE verification and display the result.

.SH OPTIONS
.PP

.TP 3
.B -v \fIlevel\fR
.EX
0 - Normal
1 - Display DANE Verification Result
2 - Everything including SMTP conversation
.EE

.TP 3
.B -m
Do a MX query for host \fIhost\fR and then query the MX hosts for TLSA Resource Records

.TP 3
.B -s
Issue STARTTLS command and attemtp DANE verification

.SH EXAMPLE USAGE
.EX
Example 1
$ dnstlsarr postino.cesnet.cz
terenasslca3ta.cesnet.cz ttl=3282 2 0 1 be6a0d9e1d115f2293f6abf11b3ec8e882e24426eeeb09aaa503597993e77a25
terenasslca3ta.cesnet.cz ttl=3282 2 0 1 beb8efe9b1a73c841b375a90e5fff8048848e3a2af66f6c4dd7b938d6fe8c5d8

Example 2 
$ dnstlsarr -v 2 -s mail.ietf.org
checking mail.ietf.org
TLSARR[0]:_25._tcp.mail.ietf.org IN TLSA ( 3 1 1 0c72ac70b745ac19998811b131d662c9ac69dbdbe7cb23e5b514b56664c5d3d6 )
220 ietfa.amsl.com ESMTP
Client: EHLO argos
250-ietfa.amsl.com
250-PIPELINING
250-SIZE 67108864
250-ETRN
250-STARTTLS
250-AUTH PLAIN LOGIN
250-AUTH=PLAIN LOGIN
250-ENHANCEDSTATUSCODES
250 8BITMIME
Client: STARTTLS
220 2.0.0 Ready to start TLS
matched sha256 fingerprint [0c72ac70b745ac19998811b131d662c9ac69dbdbe7cb23e5b514b56664c5d3d6] of subjectPublicKeyInfo
Client: QUIT
221 2.0.0 Bye

Example 3 - querying the MX record to get the TLSA RR
$ dnstlsarr -v 2 -s postino.cesnet.cz
checking postino.cesnet.cz
TLSARR[0]:terenasslca3ta.cesnet.cz IN TLSA ( 2 0 1 be6a0d9e1d115f2293f6abf11b3ec8e882e24426eeeb09aaa503597993e77a25 )
TLSARR[1]:terenasslca3ta.cesnet.cz IN TLSA ( 2 0 1 beb8efe9b1a73c841b375a90e5fff8048848e3a2af66f6c4dd7b938d6fe8c5d8 )
220 postino.cesnet.cz ESMTP
Client: EHLO argos
250-postino.cesnet.cz
250-PIPELINING
250-SIZE 41943040
250-VRFY
250-ETRN
250-STARTTLS
250-ENHANCEDSTATUSCODES
250-8BITMIME
250 DSN
Client: STARTTLS
220 2.0.0 Ready to start TLS
failed  sha256 fingerprint [be6a0d9e1d115f2293f6abf11b3ec8e882e24426eeeb09aaa503597993e77a25] of full certificate
matched sha256 fingerprint [beb8efe9b1a73c841b375a90e5fff8048848e3a2af66f6c4dd7b938d6fe8c5d8] of full certificate
Client: QUIT
221 2.0.0 Bye

Example 4 - getting TLSA RR by giving the domain name
$ dnstlsarr -m cesnet.cz
MX postino.cesnet.cz IPv6 2001:718:1:101::144:24210
terenasslca3ta.cesnet.cz ttl=2974 2 0 1 be6a0d9e1d115f2293f6abf11b3ec8e882e24426eeeb09aaa503597993e77a25
terenasslca3ta.cesnet.cz ttl=2974 2 0 1 beb8efe9b1a73c841b375a90e5fff8048848e3a2af66f6c4dd7b938d6fe8c5d8
MX postino.cesnet.cz IPv4 195.113.144.24210
terenasslca3ta.cesnet.cz ttl=2974 2 0 1 be6a0d9e1d115f2293f6abf11b3ec8e882e24426eeeb09aaa503597993e77a25
terenasslca3ta.cesnet.cz ttl=2974 2 0 1 beb8efe9b1a73c841b375a90e5fff8048848e3a2af66f6c4dd7b938d6fe8c5d8
MX cartero.cesnet.cz IPv6 2001:718:ff05:202::1650
terenasslca3ta.cesnet.cz ttl=2974 2 0 1 be6a0d9e1d115f2293f6abf11b3ec8e882e24426eeeb09aaa503597993e77a25
terenasslca3ta.cesnet.cz ttl=2974 2 0 1 beb8efe9b1a73c841b375a90e5fff8048848e3a2af66f6c4dd7b938d6fe8c5d8
MX cartero.cesnet.cz IPv4 78.128.216.1650
terenasslca3ta.cesnet.cz ttl=2974 2 0 1 be6a0d9e1d115f2293f6abf11b3ec8e882e24426eeeb09aaa503597993e77a25
terenasslca3ta.cesnet.cz ttl=2974 2 0 1 beb8efe9b1a73c841b375a90e5fff8048848e3a2af66f6c4dd7b938d6fe8c5d8
MX mail.cesnet.cz IPv4 195.113.144.234100
terenasslca3ta.cesnet.cz ttl=2974 2 0 1 be6a0d9e1d115f2293f6abf11b3ec8e882e24426eeeb09aaa503597993e77a25
terenasslca3ta.cesnet.cz ttl=2974 2 0 1 beb8efe9b1a73c841b375a90e5fff8048848e3a2af66f6c4dd7b938d6fe8c5d8

.EE

.SH RETURN VALUE
  0 for success
  1 for failure in getting TLSA Resource Records or DANE verification failure (-s option)

.SH "SEE ALSO"
qdane(8),
qmail-daned(8),
tlsacheck(3),
qmail-remote(8)


libdkim-x/NEWS
Migrated to libtool 2.x


indimail-mta-x/dnstlsarr.c
/*
 * $Log: dnstlsarr.c,v $
 * Revision 1.15  2022-08-21 17:59:08+05:30  Cprogrammer
 * fix compilation error when TLS is not defined in conf-tls
 *
 * Revision 1.14  2021-05-26 11:05:43+05:30  Cprogrammer
 * use starttls.h for prototypes in starttls.c
 *
 * Revision 1.13  2020-11-24 13:45:02+05:30  Cprogrammer
 * removed exit.h
 *
 * Revision 1.12  2018-06-01 22:52:11+05:30  Cprogrammer
 * exit on incorrect usage
 *
 * Revision 1.11  2018-06-01 16:29:52+05:30  Cprogrammer
 * added verbose messages to indicate the type of dns query
 *
 * Revision 1.10  2018-05-31 19:40:46+05:30  Cprogrammer
 * fixed compiler 'not reached' warning
 *
 * Revision 1.9  2018-05-31 14:41:33+05:30  Cprogrammer
 * included hastlsa.h
 *
 * Revision 1.8  2018-05-31 02:20:55+05:30  Cprogrammer
 * added option to query mx records before fetching TLSA Resource Records
 *
 * Revision 1.7  2018-05-30 20:17:38+05:30  Cprogrammer
 * added prototype for scan_int()
 *
 * Revision 1.6  2018-05-30 20:10:54+05:30  Cprogrammer
 * added options to do complete DANE verification
 *
 * Revision 1.5  2018-05-28 19:56:56+05:30  Cprogrammer
 * exit with 111 for wrong usage
 *
 * Revision 1.4  2018-05-26 19:17:35+05:30  Cprogrammer
 * fixed program name in usage
 *
 * Revision 1.3  2018-05-26 19:08:41+05:30  Cprogrammer
 * removed -m option to query mx records
 *
 * Revision 1.2  2018-05-26 16:10:14+05:30  Cprogrammer
 * removed leftover debugging code
 *
 * Revision 1.1  2018-05-26 12:37:25+05:30  Cprogrammer
 * Initial revision
 *
 */
#include "substdio.h"
#include "subfd.h"
#include "hastlsa.h"
#if defined(HASTLSA) && defined(TLS)
#include <sys/types.h>
#include <sys/socket.h>
#include <unistd.h>
#include "stralloc.h"
#include "fmt.h"
#include "dns.h"
#include "dnsdoe.h"
#include "tlsarralloc.h"
#include "sgetopt.h"
#include "tls.h"
#include "control.h"
#include "scan.h"
#include "now.h"
#include "starttls.h"

char            temp[IPFMT + FMT_ULONG];
int             timeoutssl = 300;
int             timeoutconnect = 60;
int             verbose;
stralloc        helohost = { 0 };
stralloc        sahost = { 0 };

extern stralloc sa;
extern tlsarralloc ta;
extern ipalloc  ia;

void
pusage()
{
	substdio_puts(subfderr,      "usage: dnstlsarr [-p port] [-c timeoutc] [-t timeoutr] [host\n");
	substdio_putsflush(subfderr, "         -p port  - port to connect to\n");
	substdio_putsflush(subfderr, "         -c timoutc - Timeout for connection to remote\n");
	substdio_putsflush(subfderr, "         -t timoutr - Timeout for data from remote\n");
	substdio_putsflush(subfderr, "         -v level - verbosity level for STARTTLS\n");
	substdio_putsflush(subfderr, "         -s Initiate STARTTLS to initiate DANE verification\n");
	_exit (111);
}

int
main(argc, argv)
	int             argc;
	char          **argv;
{
	int             opt, k, j, i, query_mx = 0, verify = 0;
	char           *port = "25", *host = (char *) 0;
	char            hex[2];
	unsigned long   r;

	while ((opt = getopt(argc,argv,"msv:p:c:t:")) != opteof) {
    	switch(opt)
		{
		case 'p':
			port = optarg;
			break;
		case 'm':
			query_mx = 1;
			break;
		case 's':
			verify = 1;
			break;
		case 'c': /*- timeoutconnect */
			scan_int(optarg, &timeoutconnect);
			break;
		case 't': /*- timeoutconnect */
			scan_int(optarg, &timeoutssl);
			break;
		case 'v':
			verbose = *optarg - '0';
			break;
		default:
			pusage();
		}
	}
	if (optind < argc)
		host = argv[optind++];
	else
		pusage();
	if (verify) { /*- DANE Verification */
		if (control_init() == -1)
			die_control("unable to read control file ", "me");
		if (control_rldef(&helohost, "helohost", 1, (char *) 0) != 1)
			die_control("unable to read control file ", "helohost");
		if (query_mx) {
			if (!stralloc_copys(&sahost, host)) {
				substdio_putsflush(subfderr, "out of memory\n");
				_exit(111);
			}
			dns_init(0);
			r = now() + getpid();
			if (verbose) {
				out("MX query for ");
				out(host);
				out("\n");
			}
			dnsdoe(dns_mxip(&ia, &sahost, r));
			for (j = 0; j < ia.len; ++j) {
				out("checking MX host ");
				out(ia.ix[j].fqdn);
				out("\n");
				flush();
				if (!(i = do_dane_validation(ia.ix[j].fqdn, 25)))
					return (i);
			}
			return (1);
		} else {
			out("checking ");
			out(host);
			out("\n");
			flush();
			return (do_dane_validation(host, 25));
		}
	}
	/*- DANE TLSA RR display mode */
	dns_init(0);
	if (query_mx) {
		if (!stralloc_copys(&sahost, host)) {
			substdio_putsflush(subfderr, "out of memory\n");
			_exit(111);
		}
		r = now() + getpid();
		if (verbose) {
			out("MX query for ");
			out(host);
			out("\n");
		}
		dnsdoe(dns_mxip(&ia, &sahost, r));
		for (k = 0; k < ia.len; ++k) {
			if (!stralloc_copyb(&sa, "_", 1))
				die_nomem();
			if (!stralloc_cats(&sa, port))
				die_nomem();
			if (!stralloc_catb(&sa, "._tcp.", 6))
				die_nomem();
			if (!stralloc_cats(&sa, ia.ix[k].fqdn))
				die_nomem();
			out("MX ");
			out(ia.ix[k].fqdn);
			switch(ia.ix[k].af)
			{
			case AF_INET:
				out(" IPv4 ");
				substdio_put(subfdout, temp, ip4_fmt(temp, &ia.ix[k].addr.ip));
			break;
#ifdef IPV6
			case AF_INET6:
				out(" IPv6 ");
				substdio_put(subfdout, temp, ip6_fmt(temp, &ia.ix[k].addr.ip6));
			break;
#endif
			default:
				substdio_puts(subfdout, "Unknown address family = ");
				substdio_put(subfdout, temp, fmt_ulong(temp, ia.ix[k].af));
			}
			substdio_put(subfdout, temp, fmt_ulong(temp, (unsigned long) ia.ix[k].pref));
			out("\n");
			if (verbose) {
				out("TLSA RR query for ");
				substdio_put(subfdout, sa.s, sa.len);
				out("\n");
			}
			dnsdoe(dns_tlsarr(&ta, &sa));
			for (j = 0; j < ta.len; ++j) {
				substdio_put(subfdout, ta.rr[j].host, ta.rr[j].hostlen);
				substdio_put(subfdout, " ttl=", 5);
				substdio_put(subfdout, temp, fmt_ulong(temp, ta.rr[j].ttl));
				substdio_put(subfdout, " ", 1);
				substdio_put(subfdout, temp, fmt_ulong(temp, (unsigned long) ta.rr[j].usage));
				substdio_put(subfdout, " ", 1);
				substdio_put(subfdout, temp, fmt_ulong(temp, (unsigned long) ta.rr[j].selector));
				substdio_put(subfdout, " ", 1);
				substdio_put(subfdout, temp, fmt_ulong(temp, (unsigned long) ta.rr[j].mtype));
				substdio_put(subfdout, " ", 1);
				for (i = 0; i < ta.rr[j].data_len; i++) {
					fmt_hexbyte(hex, (ta.rr[j].data + i)[0]);
					substdio_put(subfdout, hex, 2);
				}
				substdio_putsflush(subfdout, "\n");
			}
		}
	} else {
		if (!stralloc_copyb(&sa, "_", 1))
			die_nomem();
		if (!stralloc_cats(&sa, port))
			die_nomem();
		if (!stralloc_catb(&sa, "._tcp.", 6))
			die_nomem();
		if (!stralloc_cats(&sa, host))
			die_nomem();
		if (verbose) {
			out("TLSA RR query for ");
			substdio_put(subfdout, sa.s, sa.len);
			out("\n");
		}
		dnsdoe(dns_tlsarr(&ta, &sa));
		for (j = 0; j < ta.len; ++j) {
			substdio_put(subfdout, ta.rr[j].host, ta.rr[j].hostlen);
			substdio_put(subfdout, " ttl=", 5);
			substdio_put(subfdout, temp, fmt_ulong(temp, ta.rr[j].ttl));
			substdio_put(subfdout, " ", 1);
			substdio_put(subfdout, temp, fmt_ulong(temp, (unsigned long) ta.rr[j].usage));
			substdio_put(subfdout, " ", 1);
			substdio_put(subfdout, temp, fmt_ulong(temp, (unsigned long) ta.rr[j].selector));
			substdio_put(subfdout, " ", 1);
			substdio_put(subfdout, temp, fmt_ulong(temp, (unsigned long) ta.rr[j].mtype));
			substdio_put(subfdout, " ", 1);
			for (i = 0; i < ta.rr[j].data_len; i++) {
				fmt_hexbyte(hex, (ta.rr[j].data + i)[0]);
				substdio_put(subfdout, hex, 2);
			}
			substdio_putsflush(subfdout, "\n");
		}
	}
	flush();
	_exit(0);
	/*- Not reached */
	return (0);
}
#else
#warning "not compiled with -DHASTLSA -DTLS"
#include <unistd.h>
int
main()
{
	substdio_puts(subfderr, "not compiled with -DHASTLSA & -DTLS. Check conf-tlsa, conf-tls\n");
	substdio_flush(subfderr);
	_exit (100);
}
#endif

void
getversion_dnstlsarr_c()
{
	static char    *x = "$Id: dnstlsarr.c,v 1.15 2022-08-21 17:59:08+05:30 Cprogrammer Exp mbhangui $";

#if defined(HASTLSA) && defined(TLS)
	x = sccsidstarttlsh;
#endif
	x++;
}
